renderdoc/driver/ihv/nv/nv_counters.cpp

```diff
@@ -0,0 +1,380 @@
+                                                      &internalIds = mInternalIds
+  ](GPUCounter counter) {
+    const uint32_t externalId =
+        static_cast<uint32_t>(counter) - static_cast<uint32_t>(GPUCounter::FirstNvidia);
+    selExternalIds.push_back(counter);
+    selInternalIds.push_back(internalIds[externalId]);
+  });
+
+  if(NvPmResultFails(
+         mNvPmApi->AddCounters(mNvPmCtx, static_cast<NVPMUINT>(mSelectedInternalIds.size()),
+                               mSelectedInternalIds.data()),
+         "call to 'NvPmApi::AddCounters'"))
+  {
+    return false;
+  }
+
+  if(mObjectsCount != objectsCount)
+  {
+    if(mObjectsCount != 0)
+    {
+      NvPmResultFails(mNvPmApi->DeleteObjects(mNvPmCtx), "call to 'NvPmApi::DeleteObjects'");
+      mObjectsCount = 0;
+    }
+
+    if(NvPmResultFails(mNvPmApi->ReserveObjects(mNvPmCtx, objectsCount),
+                       "call to 'NvPmApi::ReserveObjects'"))
+    {
+      return false;
+    }
+    mObjectsCount = objectsCount;
+  }
+  return true;
+}
+
+uint32_t NVCounters::BeginExperiment() const
+{
+  NVPMUINT NumPasses = 0;
+  if(NvPmResultFails(mNvPmApi->BeginExperiment(mNvPmCtx, &NumPasses),
+                     "call to 'NvPmApi::BeginExperiment'"))
+  {
+    return 0;
+  }
+  return NumPasses;
+}
+
+void NVCounters::EndExperiment(const std::vector<uint32_t> &eventIds,
+                               std::vector<CounterResult> &Result) const
+{
+  NvPmResultFails(mNvPmApi->EndExperiment(mNvPmCtx), "call to 'NvPmApi::EndExperiment'");
+
+  // NVPMUINT NumCounters = mObjectsCount;
+  // NVPMRESULT result = mNvPmApi->SampleEx(mNvPmCtx, Samples.data(), &NumCounters);
+  // NvPmResultFails(result, "call to 'NvPmApi::SampleEx'");
+  // mNvPmApi->GetCounterValue();
+
+  Result.reserve(mSelectedExternalIds.size() * mObjectsCount);
+
+  for(uint32_t counterIdx = 0; counterIdx < mSelectedExternalIds.size(); ++counterIdx)
+  {
+    const GPUCounter counter = mSelectedExternalIds[counterIdx];
+
+    const uint32_t externalId =
+        static_cast<uint32_t>(counter) - static_cast<uint32_t>(GPUCounter::FirstNvidia);
+
+    const NVPMCounterID internalId = mInternalIds[externalId];
+
+    const uint32_t internalDesc = mInternalDescriptors[externalId];
+
+    const NVPMCOUNTERDISPLAY displayType = static_cast<NVPMCOUNTERDISPLAY>(internalDesc >> 1);
+    const NVPMCOUNTERTYPE counterType = static_cast<NVPMCOUNTERTYPE>(internalDesc & 1);
+
+    if(counterType == NVPM_VALUE_TYPE_UINT64)
+    {
+      if(displayType == NVPM_CD_RATIO)
+      {
+        for(uint32_t i = 0; i < mObjectsCount; ++i)
+        {
+          NVPMUINT64 Value;
+          NVPMUINT64 Cycles;
+          NVPMUINT8 Overflow;
+          NVPMRESULT result =
+              mNvPmApi->GetCounterValueUint64(mNvPmCtx, internalId, i, &Value, &Cycles, &Overflow);
+
+          double Ratio = static_cast<double>(Value) / static_cast<double>(Cycles);
```

Comment:
I don't have the values in front of me so I don't really know, but is it possible it would be better to report these as percentages instead of a unit-less ratio? Since "0.6" is maybe less obvious than "60%" if the value is e.g. what proportion of the time was spent doing X. Since it's always a ratio compared to cycles this feels like maybe a better fit, but you can tell better than me.

Reply:
Unfortunately, information about units for counters is almost non-existing thing. `NvPmApi` supports `float64` / `uint64` counters with display types `raw` / `ratio`. `ratio`-counters are usually defined by two `uint64` variables and it's usually something like `metric/frequency`. When the actual value for `ratio`-counter is computed, it can be a ratio (instructions per shader invocation, for example), a percentage (l2 hit rate) or a normalized ratio. The same thing is true for `raw` `float64` counters. Real meaning could be understood by a user from a description string.

Reply:
Yeh that makes sense, I hoped that maybe it would be somewhat consistent but I understand the problem.